libvisual-plugins/plugins/actor/gstreamer/actor_gstreamer.c

```diff
@@ -90,26 +91,41 @@ static int act_gstreamer_init (VisPluginData *plugin)
 
     gst_init (NULL, NULL);
 
-    char *launch_str = g_strdup_printf ("filesrc location=%s ! decodebin ! ffmpegcolorspace ! "
+    // Regarding the pipeline below:
+    // - element "filesrc" reads a video file from disk
+    // - element "decodebin" detects the video file's encoding
+    //   and serves raw pixel data
+    // - element "videoconvert" does colorspace conversion from given to wanted
+    // - element "videoscale" does scaling from given to wanted resolution
+    // - element "capsfilter" limits the output to the width/height/format we want
+    // - element "fakesink" allows accessing the rendered raw pixels from C code
+    // - "name=capsfilter" and "name=sink" are needed to access these elements
+    //   from within C code further down
+    // - "signal-handoffs=true" makes fakesink let use know it started using
+    //   a different buffer
+    // - "sync=true" makes fakesink respect the target framerate of the video
+    //   input to give the appearance of original playback speed
+    //   (in contrast to serving frames as fast as possible)
+    char *launch_str = g_strdup_printf ("filesrc location=%s ! decodebin ! videoconvert ! "
                                         "videoscale ! capsfilter name=capsfilter ! "
                                         "fakesink name=sink signal-handoffs=true sync=true",
                                         "test.mpg");
 
     GError *error = NULL;
     priv->pipeline = gst_parse_launch (launch_str, &error);
-    g_free (launch_str);
 
     if (!priv->pipeline) {
         visual_log (VISUAL_LOG_ERROR, "Failed to create pipeline: %s", error->message);
         g_error_free (error);
+        g_free (launch_str);
```

Comment:
There's a leak here when the branch is not taken i.e. pipeline is successfully constructed.

Reply:
@kaixiong oh my, good catch! :+1: